spacy/ml/tb_framework.pyx

```diff
@@ -127,72 +139,84 @@ def init(
     # TODO: Experiment with whether better to initialize upper_W
     model.set_param("lower_W", Wl)
     model.set_param("lower_b", bl)
     model.set_param("lower_pad", padl)
-    model.set_param("upper_W", Wu)
-    model.set_param("upper_b", bu)
     # model = _lsuv_init(model)
     return model
 
-def forward(model, docs_moves: Tuple[List[Doc], TransitionSystem], is_train: bool):
+InWithoutActions = Tuple[List[Doc], TransitionSystem]
+InWithActions = Tuple[List[Doc], TransitionSystem, List[Ints1d]]
+InT = TypeVar("InT", InWithoutActions, InWithActions)
+
+def forward(model, docs_moves: InT, is_train: bool):
+    if len(docs_moves) == 2:
+        docs, moves = docs_moves
+        actions = None
+    else:
+        docs, moves, actions = docs_moves
+
     beam_width = model.attrs["beam_width"]
     lower_pad = model.get_param("lower_pad")
     tok2vec = model.get_ref("tok2vec")
 
-    docs, moves = docs_moves
     states = moves.init_batch(docs)
     tokvecs, backprop_tok2vec = tok2vec(docs, is_train)
     tokvecs = model.ops.xp.vstack((tokvecs, lower_pad))
     feats, backprop_feats = _forward_precomputable_affine(model, tokvecs, is_train)
     seen_mask = _get_seen_mask(model)
 
+    # Fixme: support actions in forward_cpu
     if beam_width == 1 and not is_train and isinstance(model.ops, NumpyOps):
-        return _forward_greedy_cpu(model, moves, states, feats, seen_mask)
+        return _forward_greedy_cpu(model, moves, states, feats, seen_mask, actions=actions)
     else:
-        return _forward_fallback(model, moves, states, tokvecs, backprop_tok2vec, feats, backprop_feats, seen_mask, is_train)
+        return _forward_fallback(model, moves, states, tokvecs, backprop_tok2vec, feats, backprop_feats, seen_mask, is_train, actions)
+
 
 def _forward_greedy_cpu(model: Model, TransitionSystem moves, states: List[StateClass], np.ndarray feats,
-                        np.ndarray[np.npy_bool, ndim=1] seen_mask):
-    cdef vector[StateC *] c_states
+                np.ndarray[np.npy_bool, ndim=1] seen_mask, actions: Optional[List[Ints1d]]=None):
+    cdef vector[StateC*] c_states
     cdef StateClass state
     for state in states:
         if not state.is_final():
             c_states.push_back(state.c)
-    weights = get_c_weights(model, <float *> feats.data, seen_mask)
+    cdef object featsp = feats
+    weights = get_c_weights(model, <float*>feats.data, seen_mask)
     # Precomputed features have rows for each token, plus one for padding.
     cdef int n_tokens = feats.shape[0] - 1
     sizes = get_c_sizes(model, c_states.size(), n_tokens)
     cdef CBlas cblas = model.ops.cblas()
-    scores = _parseC(cblas, moves, &c_states[0], weights, sizes)
+    scores = _parseC(cblas, moves, &c_states[0], weights, sizes, actions=actions)
 
     def backprop(dY):
         raise ValueError(Errors.E1038)
 
     return (states, scores), backprop
 
 cdef list _parseC(CBlas cblas, TransitionSystem moves, StateC** states,
```

Comment:
Not a part of this PR per-se, but this method seems like it can benefit from a more descriptive name.

Reply:
Agreed.